Negative genes input, neutral genes output, CI deploy refactor, and other frontend additions

vincerubinetti opened the pull request:
@ChristopherMancuso please look at the negative genes input and neutral gene results portions. You'll have to run the stack locally.

@falquaddoomi please take a look at the deployment workflows. I haven't tested these on a fork yet, and haven't generated the secrets yet either.

- format workflow files with prettier
- CI deploy the backend and frontend in sequence, in single workflow, so that they are deployed close together in time. make frontend deployment dependent on function deploy success or skip.
- add link to old geneplexus
- add "negatives" to analysis input type, and "neutral gene info" to analysis output type
- tweak select component prop names and styles, and update in-situ usage of them to match
- tweak table component. pass row to custom render func, fix expansion css, more descriptive filter labels, fix filter bug
- add note about old geneplexus to about page and home page
- add "neutrals" tab in analysis results page with basic table view of neutral gene info
- add negative genes input text box on new analysis page, and incorporate into "check genes" functionality
- add redirect link on 404 page to old site
- update cloud functions to take negatives as input and return neutral gene info as output

falquaddoomi commented on .github/workflows/deploy.yaml:
```diff
@@ -0,0 +1,92 @@
+  deploy-convert-ids:
+    needs: path-changes
+    if: ${{ needs.path-changes.outputs.convert_ids == 'true' }}
+    uses: ./.github/workflows/helper-deploy-func.yaml
```
```suggestion
    uses: ./.github/workflows/deploy-func.yaml
```

falquaddoomi commented on .github/workflows/deploy.yaml:
```diff
@@ -0,0 +1,92 @@
+    secrets: inherit
+
+  deploy-frontend:
+    needs:
+      - deploy-convert-ids
+      - deploy-ml
```
Neat; I was concerned that the `deploy-frontend` step might get skipped if either of these didn't run, but it seems that the `if` statement takes priority and explicitly checks for whether they're either successful or skipped. Nice solution!

falquaddoomi commented on frontend/.env:
```diff
@@ -1,4 +1,5 @@
 VITE_TITLE=GenePlexus
 VITE_DESCRIPTION=GenePlexus enables researchers to predict novel genes similar to genes of interest based on their patterns of connectivity in human genome-scale networks.
-VITE_URL=
+VITE_URL=https://geneplexus.net
```
For what it's worth, I'm still waiting to get control over `geneplexus.net`. I don't know if you want to soft-launch on a different domain, or just optimistically change it and hope that we get control soon.

vincerubinetti replied:
Just wanted to put the variable in place so we can easily change it when we do the switch. No one's looking at the new site yet so should be okay to have fake values in there. If google indexes the netlify site right now it'll see a mismatch but that's okay.

falquaddoomi commented on .github/workflows/deploy-func.yaml:
```diff
@@ -35,77 +34,78 @@ on:
-      # and skip the download step.
-      # the hash of the current datafile is included as part of the key, so
-      # that fresh data will be fetched if the datafile has changed in GCS.
+      # if cached files aren't present, defer caching files until end of
+      # successful run. next time workflow runs, it'll retrieve cached files
+      # from previous run and skip download step. hash of current datafile
+      # included as part of key, so fresh data will be fetched if datafile has # changed in GCS.
+
```
I guess the original was too verbose? Asking so I know what writing style you prefer.

Also, seems there was a missing newline before a `#`:
```suggestion
      # if cached files aren't present, defer caching files until end of
      # successful run. next time workflow runs, it'll retrieve cached files
      # from previous run and skip download step. hash of current datafile
      # included as part of key, so fresh data will be fetched if datafile has
      # changed in GCS.
```

vincerubinetti replied:
I did shorten it while i was reformatting things but it's completely inconsequential and subjective so I'd be happy to revert it.

falquaddoomi reviewed (commented):
Looks fine to me, but it's hard to say until we try it out. I made a few comments.

ChristopherMancuso commented:
I looked over the UI and I don't think it is a good idea for the negatives to go up top like that by the input genes. To clarify, GenePlexus is always going to find its own set of negatives, it will never solely rely on what a user uploads. The user uploaded negatives will just supplement what GenePlexus determines to be a negative. Both neutrals and negatives will not be determined in the id_conversion GCP function but the ML one.

Additionally, I still don't think it will be very clear to most users what the second selection of species means in Step 3 and that is really the major update for the web server an it's a very novel feature. 

I previously have been focused on PyGenePlexus and the GCP functions but now that I believe we have all of those functionalities updated, I'm going to mock up a few UI workflows that might highlight the interesting aspects of the method more. I don't think it will be crazy changes. I should get those done early next week and I'll share with everyone and Arjun to discuss.

vincerubinetti commented:
> I don't think it is a good idea for the negatives to go up top like that by the input genes. 

Why? Would you prefer to put the box below?

> The user uploaded negatives will just supplement what GenePlexus determines to be a negative.

I will update the tooltip text I added to reflect this.

>  I still don't think it will be very clear to most users what the second selection of species

I thought having this selection was part of a sketch you made in a Google slide?

Personally I think the two species selects may warrant a paragraph of explanation on the about page for people who don't really understand model training and predictions, like me.

I am on board with the overall ethos of making things simple and opinionated, but I think you're becoming too worried about hiding things. I view the current new analysis page as pretty simple. Having that second species dropdown takes up very little space (physically and mentally) compared to the value it provides, imo. Especially since it defaults to something sensible and common, so if it a user doesn't understand what it is, they can leave it alone. And we can encourage that by adding something like "if you're unsure, leave this alone" or "(recommended)" to the default "same as above" option.

As you suggested, I'd like to have a sit down with Arjun so we can reach a consensus and stick with it.

ChristopherMancuso commented:
I'm not too worried about hiding things. It is mostly I think we are underselling the main aspect of this work, which is the ability to seamlessly translate findings from one species to another. Many, many web servers are similar to the current live version of GenePlexus in that you take a gene list from a species and discover things about it in the same species. 

I agree with your comment for the most part

> Personally I think the two species selects may warrant a paragraph of explanation on the about page for people who don't really understand model training and predictions, like me

Right now I also don't think it is clear at all what is happening, and I would like to have a user be able to understand the web server can do this cross species transfer without having to go to an about page. The about page should only be if you need to dig into the finer points of the methodology. And I think that is easily done with a few tweaks to the UI layout. I'll work on the mock ups this week and send them out to everyone.

vincerubinetti commented:
@falquaddoomi I tested this out here https://github.com/vincerubinetti/test (I'll delete this repo in a few days) with the following test workflow, and it seemed to work as expected after a few minor fixes. So I'm feeling good about this working.

<details><summary>workflow</summary>

```yaml
on:
  push:
    branches:
      - main

jobs:
  path-changes:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - uses: dorny/paths-filter@v3
        id: changes
        with:
          filters: |
            folder-a:
              - "folder-a/**"
            folder-b:
              - "folder-b/**"
    outputs:
      folder-a: ${{ steps.changes.outputs.folder-a }}
      folder-b: ${{ steps.changes.outputs.folder-b }}

  job-a:
    runs-on: ubuntu-latest
    needs: path-changes
    if: ${{ needs.path-changes.outputs.folder-a == 'true' }}
    steps:
      - run: sleep 10

  job-b:
    runs-on: ubuntu-latest
    needs: path-changes
    if: ${{ needs.path-changes.outputs.folder-b == 'true' }}
    steps:
      - run: sleep 10

  deploy:
    needs:
      - job-a
      - job-b
    if: |
      always() &&
      (needs.job-a.result == 'success' || needs.job-a.result == 'skipped') &&
      (needs.job-b.result == 'success' || needs.job-b.result == 'skipped')
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Deploy to Netlify
        uses: nwtgck/actions-netlify@v3.0
        with:
          publish-dir: ./
          production-branch: main
          production-deploy: true
          deploy-message: Deploy from GitHub Actions
          enable-pull-request-comment: false
          enable-commit-comment: false
          enable-commit-status: true
          overwrites-pull-request-comment: false
          enable-github-deployment: false
          fails-without-credentials: true
        env:
          NETLIFY_AUTH_TOKEN: ${{ secrets.NETLIFY_AUTH_TOKEN }}
          NETLIFY_SITE_ID: ${{ secrets.NETLIFY_SITE_ID }}
        timeout-minutes: 1
```

</details>

vincerubinetti commented:
> I'm not too worried about hiding things. It is mostly I think we are underselling the main aspect of this work, which is the ability to seamlessly translate findings from one species to another.

Sorry, I had a mistaken assumption about what you were saying. I'm still unsure how a UI change (like a different layout or different components) will make this clearer; I think we'll inevitably need some text and maybe iconography. But I'll await your sketches to see.

For now, is there any other change we can make to this PR better? I'd like to merge what we have, with the knowledge that it will change again later.

ChristopherMancuso commented:
For this PR a few thoughts

- when I add negative genes and then select check genes, I'm not seeing a way to see how those negative genes were converted and which network they were in.
- when a user uploads negatives, I don't think the GCP function supports adding those into the model. That is a change I would have to make. This would also probably need to be displayed in the results like the input genes are. Should we work on that in this PR or move that to the next PR and keep this just UI focused for stuff before the ML function is launched?
- In this PR I think it says neutral information is returned. That is something that would need to be added to the GCP function and then be displayed in the results section. I think we should have a whole separate PR for that.

vincerubinetti commented:
> I'm not seeing a way to see how those negative genes were converted and which network they were in.

They were marked with a `Some gene (neg.)` in the first column of the table, but I just pushed a commit to make it more explicit by adding a dedicated column for which input box each gene was from.

> *I don't think the GCP function supports adding those into the model... That is something that would need to be added to the GCP function 

I've already updated the GCP function to pass/return the appropriate params to/from the package in this PR, following the package docs.

> In this PR I think it says neutral information is returned. 

I've already added a tab to the results page for the neutral info.

> This would also probably need to be displayed in the results like the input genes are. 

I just pushed a commit that adds the negative genes at the top of the results page.

ChristopherMancuso commented:
awesome, I'm glad you were able to figure out PyGenePlexus, that is a giant a mess right now. With those changes should we change https://github.com/krishnanlab/geneplexus-app-v2/tree/negative/functions#gpz-ml to update the input to include negatives and the output to include the neutral gene info? I can change change that documentation if this a good time to do it

vincerubinetti commented:
See the changes to `frontend/src/api/types.ts` and `functions/README.md` in [that last commit](https://github.com/krishnanlab/geneplexus-app-v2/pull/49/commits/b91713e1e0a636b12e95363c91787027ec6577b2). We had a bunch of duplication between those two files. The `readme.md` had the comments for each field, and `types.ts` had the actual typescript definitions, but the `readme.md` also had (almost) valid typescript definitions too. So I just moved it all to `types.ts` to remove duplication. It also helps to have the comments on the actual type definitions so I can see them when developing on variable hover.

The only difference is that javascript/typescript has no integer types (all numbers are floats), so I just added "(integer)" to the comments where appropriate. And I added comments to `negatives` and `neutral_gene_info`; please take a look at these.

I also fixed the linting test failures.

ChristopherMancuso reviewed (approved):
This looks like all the functionality is added and working correctly to me.

vincerubinetti commented:
Well I was [one period away](https://github.com/krishnanlab/geneplexus-app-v2/commit/46cacad2cf0f1bd91597269f8aa9fc5d47a539f2) from having it work on the first try.
